core/trino-spi/src/main/java/io/trino/spi/connector/SpiVersionHolder.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.connector;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.net.URL;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+// Separate class to isolate static initialization
+final class SpiVersionHolder
+{
+    private SpiVersionHolder() {}
+
+    static final String SPI_VERSION;
+
+    static {
+        try {
+            URL resource = ConnectorContext.class.getClassLoader().getResource("io/trino/spi/trino-spi-version.txt");
+            requireNonNull(resource, "version resource not found");
+            try (InputStream inputStream = resource.openStream();
```

Comment:
I would use properties file. Up to you.

Reply:
i tried that
```java
static {
        try {
            URL resource = ConnectorContext.class.getClassLoader().getResource("io/trino/spi/trino-spi-version.txt");
            requireNonNull(resource, "version resource not found");
            Properties properties = new Properties();
            try (InputStream inputStream = resource.openStream()) {
                properties.load(inputStream);
            }
            String spiVersion = properties.getProperty("trino-spi.version");
            if (spiVersion == null || spiVersion.isBlank() || !spiVersion.strip().equals(spiVersion)) {
                throw new IllegalStateException("Malformed version resource");
            }
            SPI_VERSION = spiVersion;
        }
        catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
```

and it looks equally ugly. Will keep as is.